internal/login/qr.go

```diff
@@ -0,0 +1,488 @@
+package login
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	stdhttp "net/http"
+	"strings"
+	"time"
+
+	"github.com/ayn2op/discordo/internal/config"
+	apphttp "github.com/ayn2op/discordo/internal/http"
+	"github.com/ayn2op/discordo/internal/ui"
+	"github.com/ayn2op/tview"
+	"github.com/gdamore/tcell/v2"
+	"github.com/gorilla/websocket"
+	"github.com/skip2/go-qrcode"
+)
+
+type qrLogin struct {
+	*tview.Flex
+	app         *tview.Application
+	cfg         *config.Config
+	view        *tview.TextView
+	done        func(token string, err error)
+	conn        *websocket.Conn
+	privKey     *rsa.PrivateKey
+	cancel      context.CancelFunc
+	fingerprint string
+}
+
+func newQRLogin(app *tview.Application, cfg *config.Config, done func(token string, err error)) *qrLogin {
+	view := tview.NewTextView().
+		SetDynamicColors(true).
+		SetScrollable(true).
+		SetWrap(false).
+		SetTextAlign(tview.AlignmentCenter)
+	view.Box = ui.ConfigureBox(view.Box, &cfg.Theme)
+	view.SetTitle("Login with QR")
+
+	flex := tview.NewFlex().
+		SetDirection(tview.FlexRow).
+		AddItem(view, 0, 1, true)
+
+	q := &qrLogin{
+		app:  app,
+		cfg:  cfg,
+		view: view,
+		done: done,
+	}
+	q.Flex = flex
+
+	view.SetChangedFunc(func() {
+		q.app.QueueUpdateDraw(func() {})
+	})
+
+	view.SetInputCapture(func(ev *tcell.EventKey) *tcell.EventKey {
+		if ev.Key() == tcell.KeyEsc {
+			q.stop()
+			if q.done != nil {
+				q.done("", nil)
+			}
+			return nil
+		}
+		return ev
+	})
+
+	return q
+}
+
+
+func (q *qrLogin) centerText(s string) string {
+	_, _, _, height := q.view.GetInnerRect()
+	if height == 0 {
+		height = 40
+	}
+	lines := strings.Count(s, "\n") + 1
+	padding := (height - lines) / 2
+	if padding < 0 {
+		padding = 0
+	} else if padding < 1 && height > lines {
+		padding = 1
+	}
+	return strings.Repeat("\n", padding) + s
+}
+
+func (q *qrLogin) start() {
+	ctx, cancel := context.WithCancel(context.Background())
+	q.cancel = cancel
+	go q.run(ctx)
+}
+
+func (q *qrLogin) stop() {
+	if q.cancel != nil {
+		q.cancel()
+	}
+	if q.conn != nil {
+		q.conn.Close()
+	}
+}
+
+type raHello struct {
+	Op                string `json:"op"`
+	TimeoutMs         int    `json:"timeout_ms"`
+	HeartbeatInterval int    `json:"heartbeat_interval"`
+}
+
+type raNonceProof struct {
+	Op             string `json:"op"`
+	EncryptedNonce string `json:"encrypted_nonce"`
+}
+
+type raPendingInit struct {
+	Op          string `json:"op"`
+	Fingerprint string `json:"fingerprint"`
+}
+
+type raPendingLogin struct {
+	Op     string `json:"op"`
+	Ticket string `json:"ticket"`
+}
+
+type raPendingTicket struct {
+	Op                   string `json:"op"`
+	EncryptedUserPayload string `json:"encrypted_user_payload"`
+}
```

Comment:
```suggestion
type raHello struct {
	TimeoutMs         int    `json:"timeout_ms"`
	HeartbeatInterval int    `json:"heartbeat_interval"`
}

type raNonceProof struct {
	EncryptedNonce string `json:"encrypted_nonce"`
}

type raPendingInit struct {
	Fingerprint string `json:"fingerprint"`
}

type raPendingLogin struct {
	Ticket string `json:"ticket"`
}

type raPendingTicket struct {
	EncryptedUserPayload string `json:"encrypted_user_payload"`
}
```

Op field can be removed, since it is unused.